src/SMT.hs

```diff
@@ -406,10 +409,12 @@ expToSMT2 ctx@(Ctx behvName whn) e = case e of
 
 -- | SMT2 has no support for exponentiation, but we can do some preprocessing
 --   if the RHS is concrete to provide some limited support for exponentiation
---   TODO: support any exponentiation expression where the RHS evaluates to a concrete value
 simplifyExponentiation :: Exp Integer -> Exp Integer -> Exp Integer
-simplifyExponentiation (LitInt a) (LitInt b) = LitInt (a ^ b)
-simplifyExponentiation _ _ = error "Internal Error: exponentiation is unsupported in SMT lib"
+simplifyExponentiation a b = fromMaybe (error "Internal Error: no support for symbolic exponents in SMT lib")
+                             $   [LitInt $ a' ^ b'                      | a' <- eval a,   b' <- evalb]
+                             <|> [foldr Mul (LitInt 1) (replicate b' a) | b' <- fromInteger <$> evalb]
```

Comment:
What will happen with the `fromInteger` here if the exponent is larger than the bounds of the `Int` type?

Reply:
seems like it will just return 0?

```
Prelude> let a = (2 ^ 256) :: Integer
Prelude> (fromInteger a) :: Int
0
```

Reply:
definately an edge case, but I would prefer to either keep everything as an `Integer` here or hard fail if we overflow the `Int` type.